### PR TITLE
weighted samples

### PR DESCRIPTION
**Pull request by SebastianBocquet**
Extend `fit_to_data` to accept weighted samples. Sample `weights` are optional input. Created a new loss function `WeightedMaximumLikelihoodLoss`.

### CODE REVIEW

**Comment by danielward27** [comment]
In an effort to provide more flexibility, ``fit_to_data`` now can take an arbitrary number of arrays in a new ``data`` argument. Which should mean something like this is possible:

```python
from paramax import unwrap
import equinox as eqx

def weighted_maximum_likelihood(params, static, x, weights, key=None):
    dist = unwrap(eqx.combine(params, static))
    return -(weights * dist.log_prob(x)).sum() / len(weights)

flow, losses = fit_to_data(
    train_key,
    flow,
    (x, weights),
    learning_rate=5e-3,
    max_epochs=200,
    loss_fn=weighted_maximum_likelihood,
)
```

I probably won't add in the weighted loss function itself, for a not very interesting reason:
- I think if it were included, it feels like it should be within the maximum likelihood loss as an optional argument rather than a separate class.
- But if you do this, you get an annoying scenario where you would often need to a partial application of a (non-first) positional argument (e.g. to set ``condition=None``), which I think isn't possible e.g. with ``functools.partial``, so would require wrapping the function, which is unintuitive.

Cheers for the pull request though, if you (or others) think we should add the separate loss for weighted maximum likelihood, feel free to reopen this and I'm sure we could figure something out.
